source/common/upstream/health_checker_impl.cc

```diff
@@ -586,5 +595,265 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
+  if (request_encoder_) {
+    handleFailure(FailureType::Network);
+    expect_reset_ = true;
+    request_encoder_->getStream().resetStream(Http::StreamResetReason::LocalReset);
+  }
+  client_->close();
+}
+
+bool GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::isHealthCheckSucceeded(
+    Grpc::Status::GrpcStatus grpc_status) const {
+  if (grpc_status != Grpc::Status::GrpcStatus::Ok) {
+    return false;
+  }
+
+  if (!health_check_response_ ||
+      health_check_response_->status() != grpc::health::v1::HealthCheckResponse::SERVING) {
+    return false;
+  }
+
+  return true;
+}
+
+void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::onRpcComplete(
+    Grpc::Status::GrpcStatus grpc_status, const std::string& grpc_message, bool end_stream) {
+  logHealthCheckStatus(grpc_status, grpc_message);
+  if (isHealthCheckSucceeded(grpc_status)) {
+    handleSuccess();
+  } else {
+    handleFailure(FailureType::Active);
+  }
+
+  if (end_stream) {
+    resetState();
+  } else {
+    // resetState() will be called by onResetStream().
+    expect_reset_ = true;
+    request_encoder_->getStream().resetStream(Http::StreamResetReason::LocalReset);
+  }
+
+  if (!parent_.reuse_connection_) {
+    client_->close();
+  }
+}
+
+void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::resetState() {
+  expect_reset_ = false;
+  request_encoder_ = nullptr;
+  decoder_ = Grpc::Decoder();
+  health_check_response_.reset();
+}
+
+void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::onTimeout() {
+  ENVOY_CONN_LOG(error, "connection/stream timeout health_flags={}", *client_,
+                 HostUtility::healthFlagsToString(*host_));
+  expect_reset_ = true;
+  request_encoder_->getStream().resetStream(Http::StreamResetReason::LocalReset);
+}
+
+void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::logHealthCheckStatus(
+    Grpc::Status::GrpcStatus grpc_status, const std::string& grpc_message) {
+  const char* service_status;
+  if (!health_check_response_) {
+    service_status = "rpc_error";
+  } else {
+    switch (health_check_response_->status()) {
+    case grpc::health::v1::HealthCheckResponse::SERVING:
+      service_status = "serving";
+      break;
+    case grpc::health::v1::HealthCheckResponse::NOT_SERVING:
+      service_status = "not_serving";
+      break;
+    case grpc::health::v1::HealthCheckResponse::UNKNOWN:
+      service_status = "unknown";
+      break;
+    default:
+      // Should not happen really, Protobuf should not parse undefined enums values.
+      NOT_REACHED;
+      break;
+    }
+  }
+  std::string grpc_status_message;
+  if (grpc_status != Grpc::Status::GrpcStatus::Ok && !grpc_message.empty()) {
+    grpc_status_message = fmt::format("{} ({})", grpc_status, grpc_message);
+  } else {
+    grpc_status_message = fmt::format("{}", grpc_status);
+  }
+  ENVOY_CONN_LOG(error, "hc grpc_status={} service_status={} health_flags={}", *client_,
```

Comment:
I think this should actually be debug level. (Since upstream can error and we don't want to spew logs by default).

Reply:
fixed. Just forgot to return back after testing